File: sherpa-onnx/java-api/src/main/java/com/k2fsa/sherpa/onnx/OfflineDiacritization.java
// Copyright 2026 Matias Lin

package com.k2fsa.sherpa.onnx;

public class OfflineDiacritization {
    private long ptr = 0;

    public OfflineDiacritization(OfflineDiacritizationConfig config) {
        LibraryLoader.maybeLoad();
        ptr = newFromFile(config);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid OfflineDiacritizationConfig: failed to create native OfflineDiacritization");
        }
    }

    public String addDiacritics(String text) {
        return addDiacritics(ptr, text);
    }

    @Override
    protected void finalize() throws Throwable {
        release();
    }

    // You'd better call it manually if it is not used anymore
    public void release() {
        if (this.ptr == 0) {
            return;
        }
        delete(this.ptr);
        this.ptr = 0;
    }

    private native void delete(long ptr);

    private native long newFromFile(OfflineDiacritizationConfig config);

    private native String addDiacritics(long ptr, String text);
}

File: sherpa-onnx/kotlin-api/OfflineDiacritization.kt
package com.k2fsa.sherpa.onnx

import android.content.res.AssetManager

data class OfflineDiacritizationModelConfig(
    var cattEncoder: String = "",
    var cattDecoder: String = "",
    var numThreads: Int = 1,
    var debug: Boolean = false,
    var provider: String = "cpu",
)


data class OfflineDiacritizationConfig(
    var model: OfflineDiacritizationModelConfig,
)

class OfflineDiacritization(
    assetManager: AssetManager? = null,
    config: OfflineDiacritizationConfig,
) {
    private var ptr: Long

    init {
        ptr = if (assetManager != null) {
            newFromAsset(assetManager, config)
        } else {
            newFromFile(config)
        }
        require(ptr != 0L) {
            "Invalid OfflineDiacritizationConfig: failed to create native OfflineDiacritization"
        }
    }

    protected fun finalize() {
        if (ptr != 0L) {
            delete(ptr)
            ptr = 0
        }
    }

    fun release() = finalize()

    fun addDiacritics(text: String) = addDiacritics(ptr, text)

    private external fun delete(ptr: Long)

    private external fun addDiacritics(ptr: Long, text: String): String

    private external fun newFromAsset(
        assetManager: AssetManager,
        config: OfflineDiacritizationConfig,
    ): Long

    private external fun newFromFile(
        config: OfflineDiacritizationConfig,
    ): Long

    companion object {
        init {
            System.loadLibrary("sherpa-onnx-jni")
        }
    }
}
